sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ResourceEvaluationStage.java

```diff
@@ -0,0 +1,189 @@
+        return resourceRequirement;
+    }
+
+    protected Optional<String> getTaskName() {
+        return taskName;
+    }
+
+    @Override
+    public void evaluate(
+            MesosResourcePool offerResourcePool,
+            OfferRequirement offerRequirement,
+            OfferRecommendationSlate offerRecommendationSlate) throws OfferEvaluationException {
+        ResourceRequirement resourceRequirement = getResourceRequirement();
+        Optional<MesosResource> mesosResourceOptional = offerResourcePool.consume(resourceRequirement);
+        if (!mesosResourceOptional.isPresent()) {
+            throw new OfferEvaluationException(String.format(
+                    "Failed to satisfy resource requirement: %s",
+                    TextFormat.shortDebugString(resourceRequirement.getResource())));
+        }
+
+        final MesosResource mesosResource = mesosResourceOptional.get();
+        Resource fulfilledResource = getFulfilledResource(mesosResource);
+        if (resourceRequirement.expectsResource()) {
+            logger.info("Expects Resource");
+
+            // Compute any needed resource pool consumption / release operations
+            // as well as any additional needed Mesos Operations.  In the case
+            // where a requirement has changed for an Atomic resource, no Operations
+            // can be performed because the resource is Atomic.
+            if (expectedValueChanged(mesosResource) && !mesosResource.isAtomic()) {
+                Value reserveValue = ValueUtils.subtract(resourceRequirement.getValue(), mesosResource.getValue());
+                Value unreserveValue = ValueUtils.subtract(mesosResource.getValue(), resourceRequirement.getValue());
+
+                if (ValueUtils.compare(unreserveValue, ValueUtils.getZero(unreserveValue.getType())) > 0) {
+                    logger.info("Updates reserved resource with less reservation");
+                    Resource unreserveResource = ResourceUtils.getDesiredResource(
+                            resourceRequirement.getRole(),
+                            resourceRequirement.getPrincipal(),
+                            resourceRequirement.getName(),
+                            unreserveValue);
+                    unreserveResource = ResourceUtils.setResourceId(
+                            unreserveResource, resourceRequirement.getResourceId());
+
+                    offerResourcePool.release(new MesosResource(
+                            ResourceUtils.getUnreservedResource(resourceRequirement.getName(), unreserveValue)));
+                    offerRecommendationSlate.addUnreserveRecommendation(
+                            new UnreserveOfferRecommendation(offerResourcePool.getOffer(), unreserveResource));
+                    fulfilledResource = getFulfilledResource(new MesosResource(resourceRequirement.getResource()));
+                }
+
+                if (ValueUtils.compare(reserveValue, ValueUtils.getZero(reserveValue.getType())) > 0) {
```

Comment:
Should this be an `else if` clause?  Not possible to reserve and unreserve for a single ResourceRequirement right?